File: python/common/envelope.py
"""Versioned message envelope and canonical encoding (ADR-0002), Python peer.

Byte-for-byte the same format as ``cpp/events/envelope.hpp``. A shared golden
fixture (``tests/unit/fixtures/envelope/golden.json``) holds both
implementations to it, so a change to one that is not made to the other fails a
test rather than surfacing as a corrupt recording months later.

Canonical means one value has exactly one byte representation: fixed field
order, no optional fields, fixed-width little-endian integers, length-prefixed
strings, no floating point, no map iteration order, no locale. That is the
property AEGIS-005 needs — without it, "the two runs differ" is a question about
the encoder rather than about the engine.
"""

from __future__ import annotations

import struct
from dataclasses import dataclass, field
from enum import Enum, IntEnum

from common.clock import EventTime, serialize_nanos

ENVELOPE_SCHEMA_VERSION = 1

MAX_STRING_LENGTH = 0xFFFF
MAX_PAYLOAD_LENGTH = 0xFFFFFFFF

_U16 = struct.Struct("<H")
_U32 = struct.Struct("<I")
_U64 = struct.Struct("<Q")

FIXED_PREFIX_LENGTH = 2 + 2 + 8 + 8 + 8


class MessageType(IntEnum):
    """Message types. Numbers are permanent.

    A retired type keeps its number forever: a recorded stream outlives the code
    that wrote it, so reusing a number would make old recordings decode into the
    wrong type. M0 defines one — the platform envelope. Domain types arrive with
    the exchange in M1 (1..999) and the participant in M3 (1000..1999).
    """

    UNSPECIFIED = 0


class DecodeError(Enum):
    """Why a decode failed.

    Returned as an exception payload rather than a silent None: a malformed
    message is an expected event a feed handler must count and continue past,
    and it must be able to say which kind it was.
    """

    TRUNCATED = "message ends before its declared fields do"
    UNSUPPORTED_SCHEMA_VERSION = "envelope schema version is not understood by this build"
    UNKNOWN_MESSAGE_TYPE = "message type is not defined in this build"
    LENGTH_OVERFLOW = "declared payload length exceeds the bytes supplied"
    TRAILING_BYTES = "message carries bytes beyond its declared payload"


class EnvelopeDecodeError(ValueError):
    def __init__(self, reason: DecodeError) -> None:
        super().__init__(f"{reason.name.lower()}: {reason.value}")
        self.reason = reason


class EnvelopeEncodeError(ValueError):
    """A value cannot be represented in the canonical form."""


@dataclass(frozen=True, slots=True)
class Envelope:
    """The header every AEGIS message carries."""

    sequence: int = 0
    stream_id: int = 0
    event_time: EventTime = field(default_factory=lambda: EventTime(0))
    producer_id: str = ""
    experiment_id: str = ""
    correlation_id: str = ""
    payload: bytes = b""
    message_type: MessageType = MessageType.UNSPECIFIED
    schema_version: int = ENVELOPE_SCHEMA_VERSION


def _put_string(parts: list[bytes], text: str) -> None:
    encoded = text.encode("utf-8")
    if len(encoded) > MAX_STRING_LENGTH:
        raise EnvelopeEncodeError(
            f"string field is {len(encoded)} bytes, over the {MAX_STRING_LENGTH}-byte limit; "
            "the length prefix is 16 bits and silently truncating would corrupt the frame"
        )
    parts.append(_U16.pack(len(encoded)))
    parts.append(encoded)


def encode(envelope: Envelope) -> bytes:
    """Encode into the canonical byte form."""
    if len(envelope.payload) > MAX_PAYLOAD_LENGTH:
        raise EnvelopeEncodeError(f"payload exceeds {MAX_PAYLOAD_LENGTH} bytes")

    parts: list[bytes] = [
        _U16.pack(envelope.schema_version),
        _U16.pack(int(envelope.message_type)),
        _U64.pack(envelope.sequence),
        _U64.pack(envelope.stream_id),
        # serialize_nanos refuses a MonotonicTime: its origin differs per process,
        # so a recorded one would replay to a different value.
        _U64.pack(serialize_nanos(envelope.event_time) & 0xFFFFFFFFFFFFFFFF),
    ]
    _put_string(parts, envelope.producer_id)
    _put_string(parts, envelope.experiment_id)
    _put_string(parts, envelope.correlation_id)
    parts.append(_U32.pack(len(envelope.payload)))
    parts.append(envelope.payload)
    return b"".join(parts)


def _take_string(data: bytes, offset: int) -> tuple[str, int]:
    if offset + 2 > len(data):
        raise EnvelopeDecodeError(DecodeError.TRUNCATED)
    (length,) = _U16.unpack_from(data, offset)
    offset += 2
    if offset + length > len(data):
        raise EnvelopeDecodeError(DecodeError.TRUNCATED)
    return data[offset : offset + length].decode("utf-8"), offset + length


def decode(data: bytes) -> Envelope:
    """Decode a canonical byte form, raising EnvelopeDecodeError on any problem."""
    if len(data) < FIXED_PREFIX_LENGTH:
        raise EnvelopeDecodeError(DecodeError.TRUNCATED)

    offset = 0
    (schema_version,) = _U16.unpack_from(data, offset)
    offset += 2
    if schema_version != ENVELOPE_SCHEMA_VERSION:
        # Refused, not tolerated: guessing at an unknown layout yields plausible
        # messages that are wrong, and the run that follows looks normal.
        raise EnvelopeDecodeError(DecodeError.UNSUPPORTED_SCHEMA_VERSION)

    (raw_type,) = _U16.unpack_from(data, offset)
    offset += 2
    try:
        message_type = MessageType(raw_type)
    except ValueError:
        raise EnvelopeDecodeError(DecodeError.UNKNOWN_MESSAGE_TYPE) from None

    (sequence,) = _U64.unpack_from(data, offset)
    offset += 8
    (stream_id,) = _U64.unpack_from(data, offset)
    offset += 8
    (raw_time,) = _U64.unpack_from(data, offset)
    offset += 8
    event_nanos = raw_time - 0x1_0000_0000_0000_0000 if raw_time >= 0x8000_0000_0000_0000 else raw_time

    producer_id, offset = _take_string(data, offset)
    experiment_id, offset = _take_string(data, offset)
    correlation_id, offset = _take_string(data, offset)

    if offset + 4 > len(data):
        raise EnvelopeDecodeError(DecodeError.TRUNCATED)
    (payload_length,) = _U32.unpack_from(data, offset)
    offset += 4
    if offset + payload_length > len(data):
        raise EnvelopeDecodeError(DecodeError.LENGTH_OVERFLOW)
    payload = data[offset : offset + payload_length]
    offset += payload_length

    if offset != len(data):
        # Trailing bytes mean the framing is wrong upstream; accepting them
        # silently would let a desynchronised stream keep decoding.
        raise EnvelopeDecodeError(DecodeError.TRAILING_BYTES)

    return Envelope(
        schema_version=schema_version,
        message_type=message_type,
        sequence=sequence,
        stream_id=stream_id,
        event_time=EventTime(event_nanos),
        producer_id=producer_id,
        experiment_id=experiment_id,
        correlation_id=correlation_id,
        payload=payload,
    )

File: cpp/events/envelope.cpp
#include "cpp/events/envelope.hpp"

#include <cstddef>
#include <cstdint>
#include <limits>
#include <span>
#include <stdexcept>
#include <string>
#include <string_view>
#include <vector>

namespace aegis::events {
namespace {

constexpr std::size_t kMaxStringLength = std::numeric_limits<std::uint16_t>::max();
constexpr std::string_view kHexDigits = "0123456789abcdef";

/// Append an unsigned integer little-endian.
///
/// Explicit little-endian rather than a memcpy of the native representation: a
/// recorded stream must decode identically on any machine, and "it works here"
/// is not a property of a file format.
template <typename T>
void put_unsigned(std::vector<std::byte>& out, T value) {
  for (std::size_t index = 0; index < sizeof(T); ++index) {
    out.push_back(static_cast<std::byte>((value >> (8U * index)) & 0xFFU));
  }
}

template <typename T>
T get_unsigned(std::span<const std::byte> bytes, std::size_t offset) {
  T value{0};
  for (std::size_t index = 0; index < sizeof(T); ++index) {
    value |= static_cast<T>(std::to_integer<std::uint64_t>(bytes[offset + index]) << (8U * index));
  }
  return value;
}

void put_string(std::vector<std::byte>& out, std::string_view text) {
  if (text.size() > kMaxStringLength) {
    // Throwing, unlike a decode failure: an oversized identifier is a caller
    // bug, not a data event. Truncating to fit the 16-bit prefix would corrupt
    // the frame and every message that follows it in the stream.
    throw std::length_error("envelope string field is " + std::to_string(text.size()) +
                            " bytes, over the " + std::to_string(kMaxStringLength) +
                            "-byte limit imposed by its 16-bit length prefix");
  }
  put_unsigned<std::uint16_t>(out, static_cast<std::uint16_t>(text.size()));
  for (const char character : text) {
    out.push_back(static_cast<std::byte>(static_cast<unsigned char>(character)));
  }
}

/// Reads a length-prefixed string, advancing `offset`. Returns false when the
/// buffer is too short — the caller turns that into DecodeError::kTruncated.
bool take_string(std::span<const std::byte> bytes, std::size_t& offset, std::string& out) {
  if (offset + sizeof(std::uint16_t) > bytes.size()) {
    return false;
  }
  const auto length = get_unsigned<std::uint16_t>(bytes, offset);
  offset += sizeof(std::uint16_t);
  if (offset + length > bytes.size()) {
    return false;
  }
  out.assign(length, '\0');
  for (std::size_t index = 0; index < length; ++index) {
    out.at(index) = static_cast<char>(std::to_integer<unsigned char>(bytes[offset + index]));
  }
  offset += length;
  return true;
}

}  // namespace

bool is_known_message_type(std::uint16_t value) {
  return value == static_cast<std::uint16_t>(MessageType::kUnspecified);
}

std::string_view describe(DecodeError error) {
  switch (error) {
    case DecodeError::kTruncated:
      return "message ends before its declared fields do";
    case DecodeError::kUnsupportedSchemaVersion:
      return "envelope schema version is not understood by this build";
    case DecodeError::kUnknownMessageType:
      return "message type is not defined in this build";
    case DecodeError::kLengthOverflow:
      return "declared payload length exceeds the bytes supplied";
    case DecodeError::kTrailingBytes:
      return "message carries bytes beyond its declared payload";
  }
  return "unknown decode error";
}

std::vector<std::byte> encode(const Envelope& envelope) {
  std::vector<std::byte> out;
  out.reserve(64 + envelope.payload.size());

  // Field order is fixed and total. There are no optional fields: an encoder
  // with choices produces two byte strings for one value, and a replay hash
  // then measures the encoder rather than the engine.
  put_unsigned<std::uint16_t>(out, envelope.schema_version);
  put_unsigned<std::uint16_t>(out, static_cast<std::uint16_t>(envelope.message_type));
  put_unsigned<std::uint64_t>(out, envelope.sequence);
  put_unsigned<std::uint64_t>(out, envelope.stream_id);

  // Timestamps go through serialize_nanos, which refuses a MonotonicTime at
  // compile time. Signed nanoseconds are reinterpreted as unsigned for encoding
  // so the byte pattern is defined for negative values too.
  const auto event_nanos = common::serialize_nanos(envelope.event_time);
  put_unsigned<std::uint64_t>(out, static_cast<std::uint64_t>(event_nanos));

  put_string(out, envelope.producer_id);
  put_string(out, envelope.experiment_id);
  put_string(out, envelope.correlation_id);

  put_unsigned<std::uint32_t>(out, static_cast<std::uint32_t>(envelope.payload.size()));
  out.insert(out.end(), envelope.payload.begin(), envelope.payload.end());
  return out;
}

DecodeResult decode(std::span<const std::byte> bytes) {
  constexpr std::size_t kFixedPrefix = 2 + 2 + 8 + 8 + 8;
  if (bytes.size() < kFixedPrefix) {
    return DecodeResult::failure(DecodeError::kTruncated);
  }

  Envelope envelope;
  std::size_t offset = 0;

  envelope.schema_version = get_unsigned<std::uint16_t>(bytes, offset);
  offset += sizeof(std::uint16_t);
  if (envelope.schema_version != kEnvelopeSchemaVersion) {
    // Refused, not tolerated: a build that guesses at an unknown layout produces
    // plausible messages that are wrong, and the run that follows looks normal.
    return DecodeResult::failure(DecodeError::kUnsupportedSchemaVersion);
  }

  const auto raw_type = get_unsigned<std::uint16_t>(bytes, offset);
  offset += sizeof(std::uint16_t);
  if (!is_known_message_type(raw_type)) {
    return DecodeResult::failure(DecodeError::kUnknownMessageType);
  }
  envelope.message_type = static_cast<MessageType>(raw_type);

  envelope.sequence = get_unsigned<std::uint64_t>(bytes, offset);
  offset += sizeof(std::uint64_t);
  envelope.stream_id = get_unsigned<std::uint64_t>(bytes, offset);
  offset += sizeof(std::uint64_t);
  envelope.event_time =
      common::EventTime{static_cast<common::Nanos>(get_unsigned<std::uint64_t>(bytes, offset))};
  offset += sizeof(std::uint64_t);

  if (!take_string(bytes, offset, envelope.producer_id) ||
      !take_string(bytes, offset, envelope.experiment_id) ||
      !take_string(bytes, offset, envelope.correlation_id)) {
    return DecodeResult::failure(DecodeError::kTruncated);
  }

  if (offset + sizeof(std::uint32_t) > bytes.size()) {
    return DecodeResult::failure(DecodeError::kTruncated);
  }
  const auto payload_length = get_unsigned<std::uint32_t>(bytes, offset);
  offset += sizeof(std::uint32_t);
  if (offset + payload_length > bytes.size()) {
    return DecodeResult::failure(DecodeError::kLengthOverflow);
  }
  envelope.payload.assign(bytes.begin() + static_cast<std::ptrdiff_t>(offset),
                          bytes.begin() + static_cast<std::ptrdiff_t>(offset + payload_length));
  offset += payload_length;

  if (offset != bytes.size()) {
    // Trailing bytes mean the framing is wrong somewhere upstream. Accepting
    // them silently would let a desynchronised stream keep decoding.
    return DecodeResult::failure(DecodeError::kTrailingBytes);
  }
  return DecodeResult::success(std::move(envelope));
}

std::string to_hex(std::span<const std::byte> bytes) {
  std::string hex;
  hex.reserve(bytes.size() * 2);
  for (const auto byte : bytes) {
    const auto value = std::to_integer<unsigned char>(byte);
    hex.push_back(kHexDigits.at(static_cast<std::size_t>(value) >> 4U));
    hex.push_back(kHexDigits.at(static_cast<std::size_t>(value) & 0x0FU));
  }
  return hex;
}

std::vector<std::byte> from_hex(std::string_view hex) {
  std::vector<std::byte> bytes;
  bytes.reserve(hex.size() / 2);
  for (std::size_t index = 0; index + 1 < hex.size(); index += 2) {
    const auto high = kHexDigits.find(hex.at(index));
    const auto low = kHexDigits.find(hex.at(index + 1));
    if (high == std::string_view::npos || low == std::string_view::npos) {
      return {};
    }
    bytes.push_back(static_cast<std::byte>((high << 4U) | low));
  }
  return bytes;
}

}  // namespace aegis::events
